### ISSUE TITLE
Best approach to customize changelog templates (e.g. ordering)

### ISSUE THREAD

**Issue by afuetterer** [python-semantic-release/python-semantic-release#1122]
## Question

I used a custom changelog template and want to switch to the much improved default ones, but customize them at the same time.

I would like to know what is the best approach to customize the default templates? Should I copy all templates and macros into my project and customize just the sections I want to differ from default? Or is there a way to just change one macro and keep the rest of the defaults? Sadly, I doubt it.

I would like to change:
- the order of commit types (from alphabetic to Features, Bug Fixes, ...)
- add a compare url to the heading of each tagged release
- only display the first line of the commit, not the whole description

These changes need to be applied in 
- changes.md.j2
- versioned_changes.md.j2

Can you give me a hint of how to achieve this?

**Comment by codejedi365** [comment]
> I would like to know what is the best approach to customize the default templates? Should I copy all templates and macros into my project and customize just the sections I want to differ from default? Or is there a way to just change one macro and keep the rest of the defaults? Sadly, I doubt it.

Yes, a recursive copy from the defaults folder. I have considered the individual override but I feel it is more error prone than it seems upon outset. Still will think about it for the future.

When you go to modify the macros it may be better to make your own macros file `my-macros.j2` and then just change the imports on the associated templates because as you want to consume the default templates in the future this would make it easier to drop-in and replace.

```sh
VIRTUAL_ENV=".venv"

cp -r "$VIRTUAL_ENV/lib/python3*/site-packages/semantic-release/data/templates/angular/md/ ./templates/
```

> I would like to change:
> 
> - the order of commit types (from alphabetic to Features, Bug Fixes, ...)
> 
> - add a compare url to the heading of each tagged release
> 
> - only display the first line of the commit, not the whole description

Note that 1 & 3 are planned for the future in the default template so I do have solutions for those already which I will provide below. 

> Can you give me a hint of how to achieve this?

First change to porting the default templates to user defined templates is to hardcode the changelog file name into the changelog template if you want to support an updating changelog.

```diff
diff a/templates/CHANGELOG.md.j2 b/templates/CHANGELOG.md.j2

  #}{%  set insertion_flag = ctx.changelog_insertion_flag
+ %}{%. set this_file = "CHANGELOG.md"
  %}{%  set unreleased_commits = ctx.history.unreleased | dictsort
  %}{%  set releases = ctx.history.released.values() | list
  %}{#
  #}{%  if ctx.changelog_mode == "init"
  %}{%    include ".components/changelog_init.md.j2"
  %}{#
  #}{%  elif ctx.changelog_mode == "update"
+ %}{%    set prev_changelog_file = this_file
- %}{%    set prev_changelog_file = ctx.prev_changelog_file
  %}{%    include ".components/changelog_update.md.j2"
  %}{#
  #}{%  endif
```

To accomplish goal 3, yes you should modify `changes.md.j2`, like so

**Comment by afuetterer** [comment]
Oh my god, thank you very much for this detailed answer. This is an early christmas present.
I will try the suggestions in the coming days and come back to you.

I will copy the default templates then, as suggested, and reference your project as the origin of the code.

As you pointed out, some of the changes are planned for a future release of the default templates. I will switch then, once those are available.

One more question: After version 9.10, is it mandatory to select update mode? I need to adapt to that as well.

**Comment by codejedi365** [comment]
You're welcome! Glad I could help.

> I will copy the default templates then, as suggested, and reference your project as the origin of the code.

That's not necessary in this case but thanks!

> As you pointed out, some of the changes are planned for a future release of the default templates. I will switch then, once those are available.

I wouldn't wait on me to make those available. I just wanted you to be aware of changes to the default templates that are coming.   It takes me much longer to update default changelog templates because I have to make them compatible for all parser variants and the change type ordering is not parser agnostic. I did give you the intended changes for angular so it shouldn't require any changes on your end when they are released.

> One more question: After version 9.10, is it mandatory to select update mode? I need to adapt to that as well.

It's definitely **NOT** mandatory.  I consider a change to the default implementation a major breaking change. I added the functionality to enable an update mode as it was a highly desired feature that was dropped from v7 to v8 during the conversion to jinja based templates. You can choose to switch or say in `init` mode. There's a significant write up on the changelog templates page of the documentation. It is planned for `v10` that the default mode will be switched to `update` so if you don't have it defined in your config as `init` then you could have unintended results. I believe update mode makes the most sense because then people can modify and adjust the changelog manually if the automated results didn't include all details desired to provide the end user. If PSR is in `init` mode then any changes are overwritten upon each new release.  The update mode is also better for late adoption to an existing project where a dev team might not of used commit conventions or changed their convention along its history. Init mode assumes the latest standard and configuration applies across the entire project's history all the way back to the initial commit which is just not really feasible.

**Comment by afuetterer** [comment]
I think I am almost there.

Regarding your suggestion:
```jinja
for type_ in section_heading_order if type_ in commit_objects
```

This does not seem to work. `commit_objects` is a list of tuples, right? So I cannot easily test if a string is part of `commit_objects` using `in`.

I am struggling to wrap my head around this. Any ideas?

**Comment by codejedi365** [comment]
And with the change from a list of tuples to a dictionary, that will change the processing of the breaking changes section. Let me go pull that conversion...

**Comment by afuetterer** [comment]
Thank you very much, I will try that. I ran exactly into the "now it does not work with a dictionary, because it expects a list of tuples" scenario. I did not want to ask again for help so soon. But now that you provided it anyway. Thanks a lot.

**Comment by codejedi365** [comment]
@afuetterer, there was a lot of little changes and it all works together which is why I forgot in the first place. Not a problem. Hopefully it works now?

**Comment by afuetterer** [comment]
I will let you know, once I got it working

**Comment by afuetterer** [comment]
@codejedi365, it took a while, but now I got it working. Thanks for all your suggestions. You can close this issue as completed, if you like. But the stalebot closed it as not planned in the meantime. Thanks again.

For future reference: My templates are located here: https://github.com/afuetterer/python-re3data/tree/main/.github/templates

**Comment by codejedi365** [comment]
@afuetterer, glad to hear! Thanks for reporting back, templates look good and I like how you added installation instructions and such. 

I know you just finished adapting the templates but just a heads up that in the next release there are some upgrades to the default templates you might want to include. 

- a new `Additional Release Information` section that is populated from parsing a `NOTICE:` git footer from commits (similar to Breaking Changes)

- a license statement added to the release notes

- a new `create_pypi_url` jinja filter 

- general error avoidance & performance within templates

**Comment by afuetterer** [comment]
Thanks a lot.

I will try the new features, once released. create_pypi_url seems pretty useful.

I ran into a couple of issues with the new default templates though. I will search the issue tracker for related findings or open new ones. For example I ran into trouble with "feat!: add new breaking feature".
